src/lib.rs

```diff
@@ -438,338 +507,521 @@ impl TsunamiBuilder {
-                    (instances, ok, id_to_name, setup_fns, usernames, pk)
                 })
             })
+            .collect();
+
+        let future = futures::future::join_all(setups)
             .map_err(Error::from)
-            .and_then(
-                move |(instances, everything_ok, id_to_name, setup_fns, usernames, private_key)| {
-                    let mut term_instances = instances.clone();
-
-                    // or_else
-                    let at_the_end = move |e| {
-                        use std::mem;
-
-                        // 5. terminate all instances
-                        if !term_instances.is_empty() {
-                            debug!(log, "terminating instances");
-                            let mut termination_req =
-                                rusoto_ec2::TerminateInstancesRequest::default();
-                            termination_req.instance_ids =
-                                mem::replace(&mut term_instances, Vec::new());
-
-                            Either::A(ec2.terminate_instances(&termination_req).map_err(
-                                move |term_e| {
-                                    warn!(
-                                        log,
-                                        "failed to terminate tsunami instances: {:?}", term_e
-                                    );
-                                    // TODO: include both errors?
-                                    e
-                                },
-                            ))
-                        } else {
-                            Either::B(futures::future::err(e))
-                        }.map(|_| ())
-                            .map_err(Error::from)
+            .and_then(move |machines| {
+                let machines: HashMap<&str, _> = machines.into_iter().fold(
+                    HashMap::new(),
+                    |mut machines, (name, machine)| {
+                        machines.entry(&name).or_insert_with(Vec::new).push(machine);
+                        machines
+                    },
+                );
+                // 4. invoke F with Machine descriptors
+                let start = time::Instant::now();
+                info!(log, "quiet before the storm");
+                f(machines)
+                    .into_future()
+                    .then(|r| r.context("tsunami main routine failed"))
+                    .map_err(move |e| {
+                        crit!(log, "main tsunami routine failed");
+                        Error::from(e)
+                    })
+                    .map(move |r| (r, start))
+            })
+            .map(move |(r, start)| {
+                info!(log, "the power of the tsunami was unleashed"; "duration" => start.elapsed().as_secs());
+                r
+            })
+            .or_else(move |e| {
+                // 5. terminate all instances
+                if !&term_instances.is_empty() {
```

Comment:
You shouldn't need the `&` here; `.` automatically does dereferencing if necessary.